File: server/src/middlewares/visitorLogger.ts
import { Request, Response, NextFunction } from 'express';
import prisma from '../utils/prisma';

// ⚡ IP Throttle Map: one log per IP per 5 minutes to prevent DB spam
const IP_THROTTLE_MS = 5 * 60 * 1000; // 5 minutes
const ipThrottleMap = new Map<string, number>();

// Clean up old throttle entries every 10 minutes to prevent memory leaks
setInterval(() => {
  const now = Date.now();
  for (const [ip, ts] of ipThrottleMap.entries()) {
    if (now - ts > IP_THROTTLE_MS) ipThrottleMap.delete(ip);
  }
}, 10 * 60 * 1000);

export const visitorLogger = (req: Request, res: Response, next: NextFunction) => {
  const path = req.path;
  const shouldSkip =
    path === '/health' ||
    path.startsWith('/socket.io') ||
    path.startsWith('/uploads/');

  if (!shouldSkip) {
    // ⚡ Capture Both: Store both the cleaned IPv4 and the raw mapped signature for the audit log
    const rawIp = (req.ip || (req.headers['x-forwarded-for'] as string) || req.socket.remoteAddress || 'unknown').split(',')[0].trim();
    const cleanIp = rawIp.startsWith('::ffff:') ? rawIp.substring(7) : rawIp;
    
    // If it's pure IPv4, force the mapped version for the "Both" view. If native IPv6, just show the one.
    const ip = !cleanIp.includes(':') ? `${cleanIp}, ::ffff:${cleanIp}` : cleanIp;
    
    const now = Date.now();
    const lastLogged = ipThrottleMap.get(ip);

    // ⚡ Only create a DB record if this IP hasn't been logged recently
    if (!lastLogged || now - lastLogged > IP_THROTTLE_MS) {
      ipThrottleMap.set(ip, now);
      // Fire-and-forget: do NOT await, do NOT block the response pipeline
      prisma.visitorLog.create({
        data: { ip, userAgent: req.headers['user-agent'], path }
      }).catch(() => { /* silently ignore log failures */ });
    }
  }

  next(); // Always call next immediately — logging is fully non-blocking
};

File: server/middlewares/visitor_logger.py
from fastapi import Request
from starlette.middleware.base import BaseHTTPMiddleware
import time
import asyncio
from utils.db import db

IP_THROTTLE_MS = 5 * 60  # 5 minutes in seconds
ip_throttle_map = {}

class VisitorLoggerMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next):
        path = request.url.path
        should_skip = (
            path == '/health' or
            path.startswith('/socket.io') or
            path.startswith('/uploads/')
        )

        if not should_skip:
            raw_ip = request.headers.get("x-forwarded-for") or (request.client.host if request.client else "unknown")
            raw_ip = raw_ip.split(",")[0].strip()
            clean_ip = raw_ip[7:] if raw_ip.startswith("::ffff:") else raw_ip
            ip = f"{clean_ip}, ::ffff:{clean_ip}" if ":" not in clean_ip else clean_ip

            now = time.time()
            last_logged = ip_throttle_map.get(ip)

            if not last_logged or now - last_logged > IP_THROTTLE_MS:
                ip_throttle_map[ip] = now
                user_agent = request.headers.get("user-agent")
                
                # Fire and forget database creation
                asyncio.create_task(db.visitorlog.create(
                    data={
                        "ip": ip,
                        "userAgent": user_agent,
                        "path": path
                    }
                ))

        response = await call_next(request)
        return response
